Parsing JSON Notices (#2506)
nasa-gcn/gcn.nasa.gov: app/routes/docs.client.samples.md

            continue
    print(message.value())
```

## Working With JSON Schema

For new missions, GCN Notices are preferably distributed in JSON format. This guide describes how to programmatically read the JSON schema.

## Parsing JSON

> Read the JSON data from [sample.schema.json](https://gcn.nasa.gov/docs/notices/schema) and [sample.example.json](https://gcn.nasa.gov/docs/notices/schema), which parses it into Python dictionaries.

Reviewer: Neither of these links are to JSON schema documents that exist.

```python
import json

# Load and parse schema and example JSON files
with open('sample.schema.json', 'r') as schema_file:
    schema = json.load(schema_file)

with open('sample.example.json', 'r') as example_file:
    example = json.load(example_file)

print('Schema:', schema)
print('Example:', example)
```

## Encoding and Decoding of Embedded Data

The following code sample shows how to encode/decode a file in Python. The `base64` package includes the methods `b64decode` and `b64encode` to make this task simple.

```python
import base64

# Parse the content of your file to a base64 encoded string:
with open("path/to/your/file", 'rb') as file:
    encoded_string = base64.b64encode(file.read())

print(encoded_string)
# b'a1512dabc1b6adb3cd1b6dcb6d4c6......'

# Decode and write the content to a local file:
with open("path/to/destination/file", 'wb') as file:
    file.write(base64.b64decode(encoded_string))

```

If you want to include a FITS file in a Notice, you add a property to your schema definition in the following format:

```python
{
    type: 'string',
    contentEncoding: 'base64',
    contentMediaType: 'image/fits',
}

> ```

Reviewer: This should be formatted as JSON, not as Python code.

            continue
    print(message.value())
```

> # Working With JSON Schema

Author: ```suggestion
## Working With JSON Schema
```

> ## Working With JSON Schema

Reviewer: Please rename this section and the previous one to make it clear that one section is dealing with decoding XML data and the other is dealing with JSON data.

> GCN Notices are distributed in JSON format. This guide describes how to programmatically read JSON schema.

Reviewer: Some GCN Notices are distributed in JSON format, but not all..

> GCN Notices are distributed in JSON format. This guide describes how to programmatically read JSON schema.

Author: ```suggestion
For new missions, GCN Notices are preferably distributed in JSON format. This guide describes how to programmatically read the JSON schema.
```

Reviewer: Why are you making suggestions on your own PR? Just make the change.

## Parsing JSON Notices

Read the JSON data from [sample.schema.json](https://gcn.nasa.gov/docs/notices/schema) and [sample.example.json](https://gcn.nasa.gov/docs/notices/schema), which parses it into Python dictionaries.

```python
import json

# Load and parse schema and example JSON files
with open('sample.schema.json', 'r') as schema_file:
    schema = json.load(schema_file)

with open('sample.example.json', 'r') as example_file:
    example = json.load(example_file)

print('Schema:', schema)
print('Example:', example)

> ```

Reviewer: Please provide sample code for parsing Kafka records in a message consuming loop.

## Encoding and Decoding of Embedded Data

> The following code sample shows how to encode/decode a file in Python. The `base64` package includes the methods `b64decode` and `b64encode` to make this task simple.

Reviewer 2: ```suggestion
The following code shows how to encode/decode a GCN Notice using the `b64decode` and `b64encode` methods from the `base64` Python package.
```

Author: I still find the previous more descriptive.

Reviewer: Provide hyperlinks to these Python methods.

```python
import base64

# Parse the content of your file to a base64 encoded string:
with open("path/to/your/file", 'rb') as file:
    encoded_string = base64.b64encode(file.read())

print(encoded_string)
# b'a1512dabc1b6adb3cd1b6dcb6d4c6......'

# Decode and write the content to a local file:
with open("path/to/destination/file", 'wb') as file:
    file.write(base64.b64decode(encoded_string))

```

> For example, if we wanted to include a FITS file in a notice, you would need to add a property to your schema definition and define the value to be the following:

Reviewer 2: ```suggestion
If you want to include a FITS file in a Notice, you add a property to your schema definition in the following format:
```

<Highlight
language="json"
code={JSON.stringify(
{
type: 'string',
contentEncoding: 'base64',
contentMediaType: 'image/fits',
},
null,
2
)}

> />

Reviewer: Please fence code using triple backticks, ```, not the Highlight component. We only use the Highlight component when the code sample is generated programmatically.

Also, this does not look like valid JSON code.

Author: Please see: https://json-schema.org/understanding-json-schema/reference/non_json_data
for valid JSON code

> 

Reviewer 3: @Vidushi-GitHub can you attach a screenshot of this code sample? Does it render correctly in the browser?

Author: It was attached above too. 
<img width="662" alt="Screenshot 2024-08-06 at 3 37 44 PM" src="https://github.com/user-attachments/assets/a62ba40e-cbab-4ff3-82fa-071cfa56db8d">

> Then in your data production pipeline, you can use the encoding steps to convert your file to a bytestring, and set the value of the property to said bytestring. See [non JSON data](https://json-schema.org/understanding-json-schema/reference/non_json_data.html) for more information.

Reviewer 2: ```suggestion
In your data production pipeline, you can use the encoding steps to convert your file to a bytestring and set the value of the property to this bytestring. See [non-JSON data](https://json-schema.org/understanding-json-schema/reference/non_json_data.html) for more information.
```
